ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,136 @@
+// a rejected promise no matter what.
+const INITIALIZE_TIMEOUT_MS = 1000 * 60 * 20;
+// While polling the get endpoint,
+const INITIAL_POLLING_DELAY_MS = 2000;
+const MAX_POLLING_DELAY_MS = 15000;
+const MAX_CREATE_COUNT = 1;
+const MAX_DELETE_COUNT = 1;
+
+export class ClusterInitializer {
+  private workspaceNamespace: string;
+  private onStatusUpdate: (ClusterStatus) => void;
+  private abortSignal?: AbortSignal;
+  private currentDelayMs = INITIAL_POLLING_DELAY_MS;
+  private createCount = 0;
+  private deleteCount = 0;
+  private isInitializing = false;
+  private initializeStartTime?: number;
+  private currentCluster?: Cluster;
+
+  constructor(options: ClusterInitializerOptions) {
+    this.workspaceNamespace = options.workspaceNamespace;
+    this.onStatusUpdate = options.onStatusUpdate ? options.onStatusUpdate : () => {};
+    this.abortSignal = options.abortSignal;
+  }
+
+  private async getCluster(): Promise<Cluster> {
+    const getClusterResponse = await clusterApi().getCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    return getClusterResponse.cluster;
+  }
+
+  private async createCluster(): Promise<Cluster> {
+    if (this.createCount >= MAX_CREATE_COUNT) {
+      throw new Error('Reached max cluster create count');
+    }
+    const createClusterResponse = await clusterApi().createCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    this.createCount++;
+    return createClusterResponse.cluster;
+  }
+
+  private async resumeCluster(): Promise<void> {
+    await notebooksClusterApi().startCluster(
+      this.currentCluster.clusterNamespace, this.currentCluster.clusterName, {signal: this.abortSignal});
+  }
+
+  private async deleteCluster(): Promise<void> {
+    if (this.deleteCount >= MAX_DELETE_COUNT) {
+      throw new Error('Reached max cluster delete count');
+    }
+    await clusterApi().deleteCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    this.deleteCount++;
+  }
+
+  private isClusterRunning(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Running;
+  }
+
+  private isClusterStopped(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Stopped;
+  }
+
+  private isClusterErrored(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Error;
+  }
+
+  async initialize(): Promise<Cluster> {
+    console.log('Initializing cluster', this.workspaceNamespace);
+    if (this.isInitializing) {
+      return Promise.reject('Initialization is already in progress');
+    }
+    this.initializeStartTime = new Date().getTime();
+
+    // Overall strategy: continue polling the get-cluster endpoint, with capped exponential backoff,
+    // until we either reach our goal state (a RUNNING cluster) or run up against the overall
+    // timeout threshold.
+    while (!this.isClusterRunning()) {
```

Comment:
One thing which isn't covered here yet is error retries. Since this is a very long polling cycle (expected to commonly be ~7-10m), would be ideal if this were resilient to the occasional transient 500. Using this package on the outgoing calls would probably suffice: https://github.com/all-of-us/workbench/blob/d86dead9ffa411924105f85b1116253d7cd53412/ui/src/app/utils/retry.tsx#L8

(e.g. only fail the polling loop if the request fails X times in a row)